### ISSUE TITLE
Solving Transformer by Hand: A Step-by-Step Math Example

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#10602]
Solving Transformer by Hand: A Step-by-Step Math Example<br>
<br>
https://ift.tt/UKLmk3C<br>
<br>
Fareed Khan<br>
<br>

<div><div><div><div><div>
<figure><figcaption>Transformer in NYC (<a rel="noopener ugc nofollow" href="https://photofunia.com/" target="_blank"><strong>created from phtofunia</strong></a>)</figcaption></figure><div><h1>Solving Transformer by Hand: A Step-by-Step Math Example</h1></div>
<p>I have already written a detailed blog on how transformers work using a very small sample of the dataset, which will be my best blog ever because it has elevated my profile and given me the motivation to write more. However, that blog is incomplete as it only covers <strong>20% of the transformer architecture</strong> and contains numerous calculation errors, as pointed out by readers. After a considerable amount of time has passed since that blog, I will be revisiting the topic in this new blog.</p>
<p>My previous blog on transformer architecture (<strong>covers only 20%</strong>):</p>
<p>I plan to explain the transformer again in the same manner as I did in my previous blog <strong>(for both coders and non-coders)</strong>, providing a complete guide with a step-by-step approach to understanding how they work.</p>
<h1>Table of Contents</h1>
<h1>Step 1 — Defining our Dataset</h1>
<p>The dataset used for creating ChatGPT is <strong>570 GB.</strong> On the other hand, for our purposes, we will be using a very small dataset to perform numerical calculations visually.</p>
<figure><figcaption>Our entire dataset containing only three sentences</figcaption></figure><p>Our entire dataset contains only three sentences, all of which are dialogues taken from a TV show. Although our dataset is cleaned, in real-world scenarios like ChatGPT creation, cleaning a 570 GB dataset requires a significant amount of effort.</p>
<h1>Step 2— Finding Vocab Size</h1>
<p>The vocabulary size determines the total number of <strong>unique words</strong> in our dataset. It can be calculated using the below formula, where <strong>N</strong> is the total number of words in our dataset.</p>
<figure><figcaption>vocab_size formula where N is total number of words</figcaption></figure><p>In order to find N, we need to break our dataset into individual words.</p>
<figure><figcaption>calculating variable <strong>N</strong></figcaption></figure><p>After obtaining N, we perform a set operation to remove duplicates, and then we can count the unique words to determine the vocabulary size.</p>
<figure><figcaption>finding vocab size</figcaption></figure><p>Therefore, the vocabulary size is <strong>23</strong>, as there are <strong>23</strong> unique words in our dataset.</p>
<h1>Step 3 — Encoding</h1>
<p>Now, we need to assign a unique number to each unique word.</p>
<figure><figcaption>encoding our unique words</figcaption></figure><p>As we have considered a single token as a single word and assigned a number to it, ChatGPT has considered a portion of a word as a single token using this formula: <code>1 Token = 0.75 Word</code></p>
<p>After encoding our entire dataset, it’s time to select our input and start working with the transformer architecture.</p>
<h1>Step 4 — Calculating Embedding</h1>
<p>Let’s select a sentence from our corpus that will be processed in our transformer architecture.</p>
<figure><figcaption>Input sentence for transformer</figcaption></figure><p>We have selected our input, and we need to find an embedding vector for it. The original paper uses a <strong>512-dimensional embedding vector</strong> for each input word.</p>
<figure><figcaption>Original Paper uses 512 dimension vector</figcaption></figure><p>Since, for our case, we need to work with a smaller dimension of embedding vector to visualize how the calculation is taking place. So, we will be using a dimension of <code>6</code> for the embedding vector.</p>
<figure><figcaption>Embedding vectors of our input</figcaption></figure><p>These values of the embedding vector are between 0 and 1 and are filled randomly in the beginning. They will later be updated as our transformer starts understanding the meanings among the words.</p>
<h1>Step 5 — Calculating Positional Embedding</h1>
<p>Now we need to find positional embeddings for our input. There are two formulas for positional embedding depending on the position of the ith value of that embedding vector for each word.</p>
<figure><figcaption>Positional Embedding formula</figcaption></figure><p>As you do know, our input sentence is <strong>“when you play the game of thrones”</strong> and the starting word is <strong>“when” </strong>with a starting index (POS) value is <code>0</code>, having a dimension (d) of <code>6</code>. For <code>i</code> from <code>0 to 5</code>, we calculate the positional embedding for our first word of the input sentence.</p>
<figure><figcaption>Positional Embedding for word: <strong>When</strong></figcaption></figure><p>Similarly, we can calculate positional embedding for all the words in our input sentence.</p>
<figure><figcaption>Calculating Positional Embeddings of our input <strong>(The calculated values are rounded)</strong></figcaption></figure><h1>Step 6 — Concatenating Positional and Word Embeddings</h1>
<p>After calculating positional embedding, we need to add word embeddings and positional embeddings.</p>
<figure><figcaption>concatenation step</figcaption></figure><p>This resultant matrix from combining both matrices (<strong>Word embedding matrix </strong>and<strong> positional embedding matrix</strong>) will be considered as an input to the encoder part.</p>
<h1>Step 7 — Multi Head Attention</h1>
<p>A multi-head attention is comprised of many single-head attentions. It is up to us how many single heads we need to combine. For example, LLaMA LLM from Meta has used 32 single heads in the encoder architecture. Below is the illustrated diagram of how a single-head attention looks like.</p>
<figure><figcaption>Single Head attention in Transformer</figcaption></figure><p>There are three inputs: <strong>query</strong>, <strong>key</strong>, and <strong>value</strong>. Each of these matrices is obtained by multiplying a different set of weights matrix from the <strong>Transpose </strong>of same matrix that we computed earlier by adding the word embedding and positional embedding matrix.</p>
<p>Let’s say, for computing the query matrix, the set of weights matrix must have the number of rows the same as the number of columns of the transpose matrix, while the columns of the weights matrix can be any; for example, we suppose <code>4</code> columns in our weights matrix. The values in the weights matrix are between <code>0 and 1</code> randomly, which will later be updated when our transformer starts learning the meaning of these words.</p>
<figure><figcaption>calculating Query matrix</figcaption></figure><p>Similarly, we can compute the <strong>key</strong> and <strong>value</strong> matrices using the same procedure, but the values in the weights matrix must be different for both.</p>
<figure><figcaption>Calculating Key and Value Matrices</figcaption></figure><p>So, after multiplying matrices, the resultant <strong>query</strong>, <strong>key</strong>, and <strong>values</strong> are obtained:</p>
<figure><figcaption>Query, Key, Value matrices</figcaption></figure><p>Now that we have all three matrices, let’s start calculating single-head attention step by step.</p>
<figure><figcaption>matrix multiplication between Query and Key</figcaption></figure><p>For scaling the resultant matrix, we have to reuse the dimension of our embedding vector, which is <code>6</code>.</p>
<figure><figcaption>scaling the resultant matrix with dimension <strong>5</strong></figcaption></figure><p>The next step of <strong>masking</strong> <strong>is</strong> <strong>optional</strong>, and we won’t be calculating it. Masking is like telling the model to focus only on what’s happened before a certain point and not peek into the future while figuring out the importance of different words in a sentence. It helps the model understand things in a step-by-step manner, without cheating by looking ahead.</p>
<p>So now we will be applying the <strong>softmax</strong> operation on our scaled resultant matrix.</p>
<figure><figcaption>Applying softmax on resultant matrix</figcaption></figure><p>Doing the final multiplication step to obtain the resultant matrix from single-head attention.</p>
<figure><figcaption>calculating the final matrix of single head attention</figcaption></figure><p>We have calculated single-head attention, while multi-head attention comprises many single-head attentions, as I stated earlier. Below is a visual of how it looks like:</p>
<figure><figcaption>Multi Head attention in Transformer</figcaption></figure><p>Each single-head attention has three inputs: <strong>query</strong>, <strong>key</strong>, and <strong>value</strong>, and each three have a different set of weights. Once all single-head attentions output their resultant matrices, they will all be concatenated, and the final concatenated matrix is once again transformed linearly by multiplying it with a set of weights matrix initialized with random values, which will later get updated when the transformer starts training.</p>
<p>Since, in our case, we are considering a single-head attention, but this is how it looks if we are working with multi-head attention.</p>
<figure><figcaption>Single Head attention vs Multi Head attention</figcaption></figure><p>In either case, whether it’s single-head or multi-head attention, the resultant matrix needs to be once again transformed linearly by multiplying a set of weights matrix.</p>
<figure><figcaption>normalizing single head attention matrix</figcaption></figure><p>Make sure the linear set of weights matrix number of columns must be equal to the matrix that we computed earlier (<strong>word embedding + positional embedding</strong>) matrix number of columns, because the next step, we will be adding the resultant normalized matrix with (<strong>word embedding + positional embedding</strong>) matrix.</p>
<figure><figcaption>Output matrix of multi head attention</figcaption></figure><p>As we have computed the resultant matrix for multi-head attention, next, we will be working on adding and normalizing step.</p>
<h1>Step 8 — Adding and Normalizing</h1>
<p>Once we obtain the resultant matrix from multi-head attention, we have to add it to our original matrix. Let’s do it first.</p>
<figure><figcaption>Adding matrices to perform add and norm step</figcaption></figure><p>To normalize the above matrix, we need to compute the mean and standard deviation row-wise for each row.</p>
<figure><figcaption>calculating meand and std.</figcaption></figure><p>we subtract each value of the matrix by the corresponding row mean and divide it by the corresponding standard deviation.</p>
<figure><figcaption>normalizing the resultant matrix</figcaption></figure><p>Adding a small value of error prevents the denominator from being zero and avoids making the entire term infinity.</p>
<h1>Step 9 — Feed Forward Network</h1>
<p>After normalizing the matrix, it will be processed through a feedforward network. We will be using a very basic network that contains only one linear layer and one ReLU activation function layer. This is how it looks like visually:</p>
<figure><figcaption>Feed Forward network comparison</figcaption></figure><p>First, we need to calculate the linear layer by multiplying our last calculated matrix with a random set of weights matrix, which will be updated when the transformer starts learning, and adding the resultant matrix to a bias matrix that also contains random values.</p>
<figure><figcaption>Calculating Linear Layer</figcaption></figure><p>After calculating the linear layer, we need to pass it through the ReLU layer and use its formula.</p>
<figure><figcaption>Calculating ReLU Layer</figcaption></figure><h1>Step 10 — Adding and Normalizing Again</h1>
<p>Once we obtain the resultant matrix from feed forward network, we have to add it to the matrix that is obtained from previous add and norm step, and then normalizing it using the row wise mean and standard deviation.</p>
<figure><figcaption>Add and Norm after Feed Forward Network</figcaption></figure><p>The output matrix of this add and norm step will serve as the query and key matrix in one of the multi-head attention mechanisms present in the decoder part, which you can easily understand by tracing outward from the add and norm to the decoder section.</p>
<h1>Step 11 — Decoder Part</h1>
<p>The good news is that up until now, we have calculated <strong>Encoder part</strong>,<strong> </strong>all the steps that we have performed, from encoding our dataset to passing our matrix through the feedforward network, are unique. It means we haven’t calculated them before. But from now on, all the upcoming steps that is the remaining architecture of the transformer (<strong>Decoder part</strong>) are going to involve similar kinds of matrix multiplications.</p>
<p>Take a look at our transformer architecture. What we have covered so far and what we have to cover yet:</p>
<figure><figcaption>Upcoming steps illustration</figcaption></figure><p>We won’t be calculating the entire decoder because most of its portion contains similar calculations to what we have already done in the encoder. Calculating the decoder in detail would only make the blog lengthy due to repetitive steps. Instead, we only need to focus on the calculations of the input and output of the decoder.</p>
<p>When training, there are two inputs to the decoder. One is from the encoder, where the output matrix of the last add and norm layer serves as the <strong>query</strong> and <strong>key</strong> for the second multi-head attention layer in the decoder part. Below is the visualization of it (from <a rel="noopener ugc nofollow" href="https://www.youtube.com/watch?v=gJ9kaJsE78k&amp;t=596s" target="_blank"><strong>batool haider</strong></a>):</p>
<figure><figcaption>Visualization is from <a rel="noopener ugc nofollow" href="https://www.youtube.com/watch?v=gJ9kaJsE78k&amp;t=596s" target="_blank"><strong>Batool Haider</strong></a></figcaption></figure><p>While the value matrix comes from the decoder after the first <strong>add and norm </strong>step.</p>
<p>The second input to the decoder is the predicted text. If you remember, our input to the encoder is <code>when you play game of thrones</code> so the input to the decoder is the predicted text, which in our case is <code>you win or you die</code> .</p>
<p>But the predicted input text needs to follow a standard wrapping of tokens that make the transformer aware of where to start and where to end.</p>
<figure><figcaption>input comparison of encoder and decoder</figcaption></figure><p>Where <code>&lt;start&gt;</code> and <code>&lt;end&gt;</code> are two new tokens being introduced. Moreover, the decoder takes one token as an input at a time. It means that <code>&lt;start&gt;</code> will be served as an input, and <code>you</code> must be the predicted text for it.</p>
<figure><figcaption>Decoder input <strong>&lt;start&gt; </strong>word</figcaption></figure><p>As we already know, these embeddings are filled with random values, which will later be updated during the training process.</p>
<p>Compute rest of the blocks in the same way that we computed earlier in the encoder part.</p>
<figure><figcaption>Calculating Decoder</figcaption></figure><p>Before diving into any further details, we need to understand what masked multi-head attention is, using a simple mathematical example.</p>
<h1>Step 12 — Understanding Mask Multi Head Attention</h1>
<p>In a Transformer, the masked multi-head attention is like a spotlight that a model uses to focus on different parts of a sentence. It’s special because it doesn’t let the model cheat by looking at words that come later in the sentence. This helps the model understand and generate sentences step by step, which is important in tasks like talking or translating words into another language.</p>
<p>Suppose we have the following input matrix, where each row represents a position in the sequence, and each column represents a feature:</p>
<figure><figcaption>inpur matrix for masked multi head attentions</figcaption></figure><p>Now, let’s understand the masked multi-head attention components having two heads:</p>
<ol>
<li>
<strong>Linear Projections (Query, Key, Value): </strong>Assume the linear projections for each head: <strong>Head 1: <em>Wq</em>1​,<em>Wk</em>1​,<em>Wv</em>1</strong>​ and <strong>Head 2: <em>Wq</em>2​,<em>Wk</em>2​,<em>Wv</em>2</strong>​</li>
<li>
<strong>Calculate Attention Scores: </strong>For each head, calculate attention scores using the dot product of Query and Key, and apply the mask to prevent attending to future positions.</li>
<li>
<strong>Apply Softmax: </strong>Apply the softmax function to obtain attention weights.</li>
<li>
<strong>Weighted Summation (Value): </strong>Multiply the attention weights by the Value to get the weighted sum for each head.</li>
<li>
<strong>Concatenate and Linear Transformation: </strong>Concatenate the outputs from both heads and apply a linear transformation.</li>
</ol>
<h2>Let’s do a simplified calculation:</h2>
<p>Assuming two conditions</p>
<ul>
<li>
<strong><em>Wq</em>1​ = <em>Wk</em>1 ​= <em>Wv</em>1 ​= <em>Wq</em>2​ = <em>Wk</em>2 ​= <em>Wv</em>2​ = <em>I</em></strong>, the identity matrix.</li>
<li><strong><em>Q</em>=<em>K</em>=<em>V</em>=Input Matrix</strong></li>
</ul>
<figure><figcaption>Mask Multi Head Attention (<strong>Two Heads</strong>)</figcaption></figure><p>The concatenation step combines the outputs from the two attention heads into a single set of information. Imagine you have two friends who each give you advice on a problem. Concatenating their advice means putting both pieces of advice together so that you have a more complete view of what they suggest. In the context of the transformer model, this step helps capture different aspects of the input data from multiple perspectives, contributing to a richer representation that the model can use for further processing.</p>
<h1>Step 13 — Calculating the Predicted Word</h1>
<p>The output matrix of the last add and norm block of the decoder must contain the same number of rows as the input matrix, while the number of columns can be any. Here, we work with <code>6</code>.</p>
<figure><figcaption>Add and Norm output of decoder</figcaption></figure><p>The last <strong>add and norm block</strong> resultant matrix of the decoder must be flattened in order to match it with a linear layer to find the predicted probability of each unique word in our dataset (corpus).</p>
<figure><figcaption>flattened the last add and norm block matrix</figcaption></figure><p>This flattened layer will be passed through a linear layer to compute the <strong>logits</strong> (scores) of each unique word in our dataset.</p>
<figure><figcaption>Calculating Logits</figcaption></figure><p>Once we obtain the logits, we can use the <strong>softmax</strong> function to normalize them and find the word that contains the highest probability.</p>
<figure><figcaption>Finding the Predicted word</figcaption></figure><p>So based on our calculations, the predicted word from the decoder is <code>you</code>.</p>
<figure><figcaption>Final output of decoder</figcaption></figure><p>This predicted word <code>you</code>, will be treated as the input word for the decoder, and this process continues until the <code>&lt;end&gt;</code> token is predicted.</p>
<h1>Important Points</h1>
<ol>
<li>The above example is very simple, as it does not involve epochs or any other important parameters that can only be visualized using a programming language like Python.</li>
<li>It has shown the process only until training, while evaluation or testing cannot be visually seen using this matrix approach.</li>
<li>Masked multi-head attentions can be used to prevent the transformer from looking at the future, helping to avoid overfitting your model.</li>
</ol>
<h1>Conclusion</h1>
<p>In this blog, I have shown you a very basic way of how transformers mathematically work using matrix approaches. We have applied positional encoding, softmax, feedforward network, and most importantly, multi-head attention.</p>
<p>In the future, I will be posting more blogs on transformers and LLM as my core focus is on NLP. More importantly, if you want to build your own million-parameter LLM from scratch using Python, I have written a blog on it which has received a lot of appreciation on Medium. You can read it here:</p>
<p><strong>Have a great time reading!</strong></p>
</div></div></div></div></div><br>
<br>
<br>
<br>
<br>
<br>
via Medium <br>
<br>
January 16, 2024 at 06:24PM
